mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+package cat
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/creachadair/taskgroup"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/mempool"
+	"github.com/tendermint/tendermint/proxy"
+	"github.com/tendermint/tendermint/types"
+)
+
+var _ mempool.Mempool = (*TxPool)(nil)
+
+const evictedTxCacheSize = 200
+
+var (
+	ErrTxInMempool       = errors.New("tx already exists in mempool")
+	ErrTxAlreadyRejected = errors.New("tx was previously rejected")
+)
+
+// TxPoolOption sets an optional parameter on the TxPool.
+type TxPoolOption func(*TxPool)
+
+// TxPool implemements the Mempool interface and allows the application to
+// set priority values on transactions in the CheckTx response. When selecting
+// transactions to include in a block, higher-priority transactions are chosen
+// first.  When evicting transactions from the mempool for size constraints,
+// lower-priority transactions are evicted sooner.
+//
+// Within the txpool, transactions are ordered by time of arrival, and are
+// gossiped to the rest of the network based on that order (gossip order does
+// not take priority into account).
+type TxPool struct {
+	// Immutable fields
+	logger       log.Logger
+	config       *config.MempoolConfig
+	proxyAppConn proxy.AppConnMempool
+	metrics      *mempool.Metrics
+
+	updateMtx            sync.Mutex
+	notifiedTxsAvailable bool
+	txsAvailable         chan struct{} // one value sent per height when mempool is not empty
+	preCheckFn           mempool.PreCheckFunc
+	postCheckFn          mempool.PostCheckFunc
+	height               int64 // the latest height passed to Update
+
+	// Thread-safe cache of rejected transactions for quick look-up
+	rejectedTxCache *LRUTxCache
+	// Thread-safe cache of valid txs that were evicted
+	evictedTxs *EvictedTxCache
+	// Thread-safe list of transactions peers have seen that we have not yet seen
+	seenByPeersSet *SeenTxSet
+
+	// Store of wrapped transactions
+	store *store
```

Comment:
is this threadsafe? or is it controlled by the `updateMtx`? 

If threadsafe, let's update the comment, otherwise let's move it to the other fields that belong to the same mutex.

Reply:
It's threadsafe. I'll add a comment